src/hammer-vlsi/hammer_vlsi.py

```diff
@@ -563,8 +679,313 @@ def logger(self, value: HammerVLSILoggingContext) -> None:
+                    break
+            assert step_id != -1
+
+            if action.location == HookLocation.ReplaceStep:
+                assert action.target_name == action.step.name, "Replacement step should have the same name"
+                new_steps[step_id] = action.step
+            elif action.location == HookLocation.InsertPreStep:
+                if has_step(action.step.name):
+                    self.logger.error("New step '{step}' already exists".format(step=action.step.name))
+                    return False
+                new_steps.insert(step_id, action.step)
+            elif action.location == HookLocation.InsertPostStep:
+                if has_step(action.step.name):
+                    self.logger.error("New step '{step}' already exists".format(step=action.step.name))
+                    return False
+                new_steps.insert(step_id + 1, action.step)
+            elif action.location == HookLocation.ResumePreStep or action.location == HookLocation.ResumePostStep:
+                if resume_step is not None:
+                    self.logger.error("More than one resume hook is present")
+                    return False
+                resume_step = action.target_name
+                resume_step_pre = action.location == HookLocation.ResumePreStep
+            else:
+                assert False, "Should not reach here"
+
+        # Check function types before running
+        for step in new_steps:
+            if not isinstance(step, HammerToolStep):
+                raise ValueError("Element in List[HammerToolStep] is not a HammerToolStep")
+            check_hammer_step_function(step.func)
+
+        # Run steps.
+        prev_step = None  # type: HammerToolStep
+
+        for step_index in range(len(new_steps)):
+            step = new_steps[step_index]
+
+            self.logger.debug("Running sub-step '{step}'".format(step=step.name))
+
+            # Do this step?
+            do_step = True
+
+            if resume_step is not None:
+                if resume_step_pre:
+                    if resume_step == step.name:
+                        self.logger.info("Resuming before '{step}' due to resume hook".format(step=step.name))
+                        # Remove resume marker
+                        resume_step = None
+                    else:
+                        self.logger.info("Sub-step '{step}' skipped due to resume hook".format(step=step.name))
+                        do_step = False
+                else:
+                    self.logger.info("Sub-step '{step}' skipped due to resume hook".format(step=step.name))
+                    do_step = False
+
+            if do_step:
+                try:
+                    if prev_step is None:
+                        # Run pre-step hook.
+                        self.do_pre_steps(step)
+                    else:
+                        # TODO: find a cleaner way of detecting a pause hook
+                        if step.name == "pause":
```

Comment:
It seems like you might not be able to have multiple pause steps since earlier you had been checking for duplicate names?

Reply:
Yeah, though the pause step is really meant to be injected once, so I'm not sure how much sense it makes to have multiple pause steps.